### PR TITLE
IHR2-2291 add documentation

### PR DESCRIPTION
**Pull request by JouriFledderman**

### CODE REVIEW

File: core/README.MD [hunk 1]
```diff
@@ -20,12 +20,70 @@ The [CoreDataFetcher](https://github.com/dotwebstack/dotwebstack-framework/blob/
...
+### 1.1.2 DelegateDataFetcher
 
 A [DelegateDataFetcher](https://github.com/dotwebstack/dotwebstack-framework/blob/IHR2-2318/core/src/main/java/org/dotwebstack/framework/core/datafetchers/DelegateDataFetcher.java) can act as a *decorator* for a *SourceDataFetcher* and gives you the capability to transform the result from a *SourceDataFetcher*
 
+
+## 1.2 Argument Constraints
+
+It is possible to add a cosntraint to an argument on any level by preceding the argument with the `@constraint` directive. 
+These constaints can be used to restrict the input from the user and also to be able give hints as to why certain arguments 
+are invalid. You can currently use one of the following constraints: `Min`, `Max`, `OneOf` and `OneOfInt`
```
Comment (msimons) [inline core/README.MD]: cosntraint=constraint & constaints=constraints

Comment (msimons) [reply core/README.MD]: Constraint argument camel-case opnemen. "min" i.p.v "Min" etc.

File: core/README.MD [hunk 2]
```diff
@@ -20,12 +20,70 @@ The [CoreDataFetcher](https://github.com/dotwebstack/dotwebstack-framework/blob/
...
+ breweries(pageSize: Int = 10 @constraint(oneOfInt: [10, 20, 50])): [Brewery!]!
+```
+
+With the `oneOfInt` constraint you can restrict an argument to be any integer within a given selection 
+of integers, but none of the integers that are not in this selection. So in this case is allowed to 
+provide an argument with a value of 10, 20 or 50 only. 
+
+## 1.3 Transform Directive
+The transform directive allows the user to transform a scalar field based on a given jexl expression. 
+This directive takes one argument `expr` that contains the jexl expression to transform the original
```
Comment (msimons) [inline core/README.MD]: 'jexl' linken aan 'http://commons.apache.org/proper/commons-jexl/'

File: backend/rdf4j/README.MD [hunk 3]
```diff
@@ -1,8 +1,112 @@
...
 
-## Field selections
+`dotwebstack.Rdf4j` is reserved by the Rdf4j backend as its properties location. The `shape.graph` property is used to 
+point to the graph that is used within the `shapes.trig` file. The `shape.prefix` property is used to point to the prefix
+of the shapesfile. This prefix is ignored when mapping GraphQL types from the `schema` to the node shapes in the
+`shapes` file. Within the `prefixes` property you can define a list of general prefixes that is added to the 
+[QueryEnvironment](https://github.com/dotwebstack/dotwebstack-framework/blob/v0.2/backend/Rdf4j/src/main/java/org/dotwebstack/framework/backend/Rdf4j/query/QueryEnvironment.java). 
 
-The RDF4J backend supports GraphQL field selections on any level within a graph. 
+Besides the configuration in the `application.yml` you need to add a `schema` file that defines the GraphQL schema.
```
Comment (bmklaijkadaster) [inline backend/rdf4j/README.MD]: 'schema.graphqls' ipv 'schema'

File: backend/rdf4j/README.MD [hunk 4]
```diff
@@ -1,8 +1,112 @@
...
+  .
+}
+```
+
+First of all, notice that the named graph in the shapes file `https://github.com/dotwebstack/beer/shapes` is the one 
+that is referred to by the `shape.graph` property in the snippet from `application.yml`. Also notice that the uri from 
+the `beer_sh:` prefix is the one that is referred to by the `shape.prefix` property in the `application.yml`. A GraphQL 
+type from `schema` is matched on a `node shape` in the `shapes` file after stripping the prefix from the name 
+of the node shape. This means that the `Brewery` from the `schema` file is mapped on the `beer_sh:Brewery` 
+node shape. Fields from `schema` file are mapped on property shapes from the `shapes` file.
```
Comment (bmklaijkadaster) [inline backend/rdf4j/README.MD]: are maped on property shapes from the 'shapes' file based on the 'sh:name' property

File: backend/rdf4j/README.MD [hunk 5]
```diff
@@ -16,15 +120,34 @@ The RDF4J backend supports GraphQL field selections on any level within a graph.
 }
 ```
 
-## Property paths
+## 1.3 Sparql directive
+
+The Rdf4j backend introduces the `@sparql` directive, this directive is added to a query in the `schema` file. It allows
```
Comment (bmklaijkadaster) [inline backend/rdf4j/README.MD]:
```suggestion
The Rdf4j backend introduces the `@sparql` directive which can be added to a query in the `schema` file.
```

File: backend/rdf4j/README.MD [hunk 6]
```diff
@@ -16,15 +120,34 @@ The RDF4J backend supports GraphQL field selections on any level within a graph.
...
+## 1.3 Sparql directive
+
+The Rdf4j backend introduces the `@sparql` directive, this directive is added to a query in the `schema` file. It allows 
+you to specify and assign paging and sort parameters and the specific query to a repository.
+
+## 1.4 Paging parameters
+
+Within the `@sparql` directive it is possible to declare a `limit` and an `offset` parameter. Arguments for both of 
+these parameters are treated as [jexl](http://commons.apache.org/proper/commons-jexl/) expressions. The results of 
+both expressions need to be an integer and it is validated that `limit > 1` and `offset > 0`. The resulting values
```
Comment (bmklaijkadaster) [inline backend/rdf4j/README.MD]:
```suggestion
Arguments for both of these parameters are treated as [jexl](http://commons.apache.org/proper/commons-jexl/) expressions and should evaluate to an integer.  The resulting limit should be greater than 1, and the resulting offset should be greater than 0.
```

File: backend/rdf4j/README.MD [hunk 7]
```diff
@@ -16,15 +120,34 @@ The RDF4J backend supports GraphQL field selections on any level within a graph.
...
+of these parameters are used as paging parameters in the resulting sparql query.  
+
+## 1.5 Sorting
+
+Within the `@sparql` directive it is possible to declare an `orderBy` parameter. The argument of this parameter needs to 
+be one or any array of objects containing a `field` and an `order` parameter. The `field` parameter needs to be
+a string and the `order` parameter is an `enum` and needs to be either `ASC` or `DESC`. The resulting values of these 
+parameters are in the order by clause in the resulting sparql query.
```
Comment (bmklaijkadaster) [inline backend/rdf4j/README.MD]:
```suggestion
parameters are used in the order by clause in the resulting sparql query.  
```

File: core/README.MD [hunk 8]
```diff
@@ -15,17 +14,110 @@ public interface CoreDataFetcher<T> extends DataFetcher<T> {
...
+The [CoreDataFetcher](https://github.com/dotwebstack/dotwebstack-framework/blob/v0.2/core/src/main/java/org/dotwebstack/framework/core/datafetchers/CoreDataFetcher.java) interface offers the capability to evaluate the support for a given [DataFetchingEnvironment](https://github.com/graphql-java/graphql-java/blob/master/src/main/java/graphql/schema/DataFetchingEnvironment.java) with the *supports* method.
+
+We are currently offering two abstract implementations of the [CoreDataFetcher](https://github.com/dotwebstack/dotwebstack-framework/blob/v0.2/core/src/main/java/org/dotwebstack/framework/core/datafetchers/CoreDataFetcher.java) which can be used by anyone. When you write your own [CoreDataFetcher](https://github.com/dotwebstack/dotwebstack-framework/blob/v0.2/core/src/main/java/org/dotwebstack/framework/core/datafetchers/CoreDataFetcher.java) and annotate it with a Spring *@Component* annotation it will be automatically registered.
```
Comment (bmklaijkadaster) [inline core/README.MD]:
```suggestion
We currently provide two abstract implementations...
```

File: core/README.MD [hunk 9]
```diff
@@ -15,17 +14,110 @@ public interface CoreDataFetcher<T> extends DataFetcher<T> {
...
+
+
+### 1.1.1 SourceDataFetcher
+
+A [SourceDataFetcher](https://github.com/dotwebstack/dotwebstack-framework/blob/v0.2/core/src/main/java/org/dotwebstack/framework/core/datafetchers/SourceDataFetcher.java) can be used to actually fetch data from *any* source.
```
Comment (bmklaijkadaster) [inline core/README.MD]: 'actually'? klinkt wat vaag, weglaten?

File: core/README.MD [hunk 10]
```diff
@@ -15,17 +14,110 @@ public interface CoreDataFetcher<T> extends DataFetcher<T> {
...
+
+### 1.1.2 DelegateDataFetcher
+
+A [DelegateDataFetcher](https://github.com/dotwebstack/dotwebstack-framework/blob/v0.2/core/src/main/java/org/dotwebstack/framework/core/datafetchers/DelegateDataFetcher.java) can act as a *decorator* for a [SourceDataFetcher](https://github.com/dotwebstack/dotwebstack-framework/blob/v0.2/core/src/main/java/org/dotwebstack/framework/core/datafetchers/SourceDataFetcher.java) and gives you the capability to transform the result from a [SourceDataFetcher](https://github.com/dotwebstack/dotwebstack-framework/blob/v0.2/core/src/main/java/org/dotwebstack/framework/core/datafetchers/SourceDataFetcher.java)
+
+
+## 1.2 Scalar Types
+Dotwebstack offers support for the *Date* and *DateTime* scalars besides the scalar types that have built-in support 
+from GraphQL *(Int, Float, String, Boolean, ID)*. Dotwebstack also offers support to write your own scalar in a
+pluggable way. To create your own Scalar Type you need to do the following:
```
Comment (bmklaijkadaster) [inline core/README.MD]:
```suggestion
It is also possible to create a custom Scalar Type:
```

File: core/README.MD [hunk 11]
```diff
@@ -15,17 +14,110 @@ public interface CoreDataFetcher<T> extends DataFetcher<T> {
...
+
+- Add the scalar to the RunTimeWiring
+```java
+  builder.scalar(CoreScalars.DATE)
+```
+
+## 1.3 Argument Constraints
+
+It is possible to add a constraint to an argument on any level by preceding the argument with the `@constraint` directive. 
+These constraints can be used to restrict the input from the user and also to be able give hints as to why certain arguments
```
Comment (bmklaijkadaster) [inline core/README.MD]:
```suggestion
These constraints can be used to restrict the input from the user and provide hints as to why certain arguments 
```

File: core/README.MD [hunk 12]
```diff
@@ -15,17 +14,110 @@ public interface CoreDataFetcher<T> extends DataFetcher<T> {
...
+- Add the scalar to the RunTimeWiring
+```java
+  builder.scalar(CoreScalars.DATE)
+```
+
+## 1.3 Argument Constraints
+
+It is possible to add a constraint to an argument on any level by preceding the argument with the `@constraint` directive. 
+These constraints can be used to restrict the input from the user and also to be able give hints as to why certain arguments 
+are invalid. You can currently use one of the following constraints: `min`, `max`, `oneOf` and `oneOfInt`
```
Comment (bmklaijkadaster) [inline core/README.MD]:
```suggestion
are invalid. The following constraints are currently provided: `min`, `max`, `oneOf` and `oneOfInt`
```

File: core/README.MD [hunk 13]
```diff
@@ -15,17 +14,110 @@ public interface CoreDataFetcher<T> extends DataFetcher<T> {
...
+it means that the pageSize provided by the user can be no higher than 100. 
+
+### 1.3.3 OneOf
+
+```graphql
+ breweries(country: String @constraint(oneOf: ["BE","DE", "EN", "NL"])): [Brewery!]!
+```
+
+With the `oneOf` constraint you can restrict an argument to match any string within a given selection 
+of strings, but none of the strings that are not in this selection. So in this case the user is allowed to
```
Comment (bmklaijkadaster) [inline core/README.MD]:
```suggestion
With the `oneOf` an argument can be restricted to match any string within the provided list.
```

File: core/README.MD [hunk 14]
```diff
@@ -15,17 +14,110 @@ public interface CoreDataFetcher<T> extends DataFetcher<T> {
...
+### 1.3.4 OneOfInt
 
-### SourceDataFetcher
- 
-A [SourceDataFetcher](https://github.com/dotwebstack/dotwebstack-framework/blob/IHR2-2318/core/src/main/java/org/dotwebstack/framework/core/datafetchers/SourceDataFetcher.java) can be used to actually fetch data from *any* source.
+```graphql
+ breweries(pageSize: Int = 10 @constraint(oneOfInt: [10, 20, 50])): [Brewery!]!
+```
 
+With the `oneOfInt` constraint you can restrict an argument to be any integer within a given selection
```
Comment (bmklaijkadaster) [inline core/README.MD]: With the `oneOfInt` constraint an argument can be restricted to match any integer within the provided list.

**Pull request merged**